### DISCUSSION TITLE
watch yields different results for same objects?

### DISCUSSION THREAD

**Discussion by echelonh** [kube-rs/kube#1256] [General]
I am attempting to write "general-purpose" watcher - I provide a configuration yaml which creates API objects:
```rust
let input = "apiVersion: apps/v1
kind: Deployment
---
apiVersion: v1
kind: Pod
"; // can contain other descriptions
for document in serde_yaml::Deserializer::from_str(input) {
    let tm = TypeMeta::deserialize(document)?;
    let gvk = GroupVersionKind::try_from(&tm)?;
    let ar = ApiResource::from_gvk(&gvk);
    let api: Api<DynamicObject> = Api::all_with(client.clone(), &ar);
    ...
    let w = watcher(api, watcher::Config::default()).touched_objects();
    while let Some(obj) = w.try_next().await? {
        debug!("{:#?}", &obj);
    }
}
```

Thing is - I'm actually interested down the road in the `kind` property of the object to be able to preceive which object/gvk I'm dealing with - and it's not always present.

iiuc - I narrowed the issue down to the difference between `list` - which does not add these properties to the object, and happens as the first stage in the state machine, vs. `watch` which does, but I wasn't able to confirm this through code.

So I wonder if this is the expected behaviour by design, a bug, or just a nuance that wasn't noticed till now and can be patched / fixed.

Thanks a lot!

**Comment by clux** [comment]
Yeah, don't think the `watch` always includes the kind in the stream, and assumes you know what you have. In general the object that you created the `Api` for does implement `Resource` and you can use this to grab the `kind` via https://docs.rs/kube/latest/kube/trait.Resource.html#tymethod.kind . For static objects where you have the struct this is easier, but since you are working with unstructured objects you need to keep the `ar` around to pass in for this (`DynamicType == ApiResource`). Here's a diff for the `dynamic_watcher` example: https://github.com/kube-rs/kube/pull/1258/files

**Reply by echelonh** [reply]
Thanks for the swift reply, much appreciated! And it's probably also the place to thank for all the incredible work you're doing here :pray: !

Indeed at some point I was just passing the original `ar` and grabbing `kind` out of it, but having seen `kind` present *sometimes* for the watched objects, I wasn't sure if I'm doing it properly.

If you don't mind explaining and know the answer - I'm still intrigued as to why do some objects appear to have `apiVersion` and `kind` fields, while others lack these?

Thanks again!

**Reply by clux** [reply]
Thank you.

> If you don't mind explaining and know the answer - I'm still intrigued as to why do some objects appear to have apiVersion and kind fields, while others lack these?

I don't remember all of this, but it falls out of the encoding from the apiserver somewhere. For some api calls the apiserver gives precise info, on others not. Don't know if it's a Kubernetes version related problem, a distribution related thing, or just a general encoding choice they've made :shrug:
